File: spring.py
#runtime - 3.763s
# uncomment the next line if running in a notebook
# %matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

# mass, spring constant, initial position and velocity
m = 1
k = 1
x = 0
v = 1

# simulation time, timestep and time
t_max = 100 #1000
dt = 0.1 #1
t_array = np.arange(0, t_max, dt)


def euler_int(m,k,x,v,dt,t_array):
    x_list = []
    v_list = []

    for t in t_array:
        # append current state to trajectories
        x_list.append(x)
        v_list.append(v)

        # calculate new position and velocity
        a = -k * x / m
        x = x + dt * v
        v = v + dt * a
    
    # convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
    x_array = np.array(x_list)
    v_array = np.array(v_list)

    # plot the position-time graph
    plt.figure(1)
    plt.clf()
    plt.xlabel('time (s)')
    plt.grid()
    plt.plot(t_array, x_array, label='x (m)')
    plt.plot(t_array, v_array, label='v (m/s)')
    plt.legend()
    plt.show()



def verlet_int(m,k,x,v,dt,t_array):
    x_list = [x-v*dt]
    v_list = []

    for t in t_array:
        # append current state to trajectories
        x_list.append(x)
        v_list.append(v)

        # calculate new position and velocity
        a = -k * x / m
        v = (1/dt) * (x - x_list[-2])
        x = 2*x -x_list[-2]+ a*(dt**2)
    
    # convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
    x_array = np.array(x_list[1:])
    v_array = np.array(v_list)

    # plot the position-time graph
    plt.figure(1)
    plt.clf()
    plt.xlabel('time (s)')
    plt.grid()
    plt.plot(t_array, x_array, label='x (m)')
    plt.plot(t_array, v_array, label='v (m/s)')
    plt.legend()
    plt.show()

#print(euler_int(m,k,x,v,dt,t_array),verlet_int(m,k,x,v,dt,t_array))
#print(verlet_int(m,k,x,v,dt,t_array))

File: spring.cpp
// runtime - 2.778s without optimizer
// runtime - 1.093s with optimizer
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int main() {

  // declare variables
  double m, k, x, v, t_max, dt, t, a;
  vector<double> t_list, x_list, v_list;

  // mass, spring constant, initial position and velocity
  m = 1;
  k = 1;
  x = 0;
  v = 1;

  // simulation time and timestep
  t_max = 100;
  dt = 0.1;

  // Euler integration
  for (t = 0; t <= t_max; t = t + dt) {

    // append current state to trajectories
    t_list.push_back(t);
    x_list.push_back(x);
    v_list.push_back(v);

    // calculate new position and velocity
    a = -k * x / m;
    x = x + dt * v;
    v = v + dt * a;

  }



  // Verlet integration
  for (t = 0; t <= t_max; t = t + dt) {

    // append current state to trajectories
    t_list.push_back(t);
    x_list.push_back(x-v*dt);
    x_list.push_back(x);
    v_list.push_back(v);

    // calculate new position and velocity
    a = -k * x / m;
    v = (0.5/dt) * (x - x_list[-2]);
    x = 2*x -x_list[-2]+ a*(pow(dt,2));

  }

  // Write the trajectories to file
  ofstream fouteuler;
  fouteuler.open("trajectories.txt");
  if (fouteuler) { // file opened successfully
    for (int i = 0; i < t_list.size(); i = i + 1) {
      fouteuler << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
    }
  } else { // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }

  ofstream verletfout;
  verletfout.open("trajectories.txt");
  if (verletfout) { // file opened successfully
    for (int i = 0; i < t_list.size(); i = i + 1) {
      verletfout << t_list[i] << ' ' << x_list[i+1] << ' ' << v_list[i] << endl;
    }
  } else { // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }

  /* The file can be loaded and visualised in Python as follows:

  import numpy as np
  import matplotlib.pyplot as plt
  results = np.loadtxt('trajectories.txt')
  plt.figure(1)
  plt.clf()
  plt.xlabel('time (s)')
  plt.grid()
  plt.plot(results[:, 0], results[:, 1], label='x (m)')
  plt.plot(results[:, 0], results[:, 2], label='v (m/s)')
  plt.legend()
  plt.show()

  */
}
